### DISCUSSION TITLE
Issue on functions with slice inputs. &[ ]

### DISCUSSION THREAD

**Discussion by RaulTrombin** [PyO3/pyo3#3329] [Questions]
![image](https://github.com/PyO3/pyo3/assets/80598030/61557dce-5f3a-4a93-bdf9-01d1c3ae0ad1)

Hi, I'm currently utilizing pyo3 for a multibinding macro, and I'm encountering difficulties with slice inputs.

Could someone provide me some guidance on how to fix or a workaround for this issue?

Out of curiosity, here is the macro: https://crates.io/crates/cpy-binder. We use it to create multibindings for both C and Python.

**Comment by adamreichold** [comment]
I think the problem you encounter is that `b"foo"` is turned into `[u8; 3]` by the Rust compiler instead of `&[u8]` and hence often needs to be coerced into a slice by e.g. `&b"foo"[..]` or similar code.

I think to really help you, it would be necessary to see the call site which produces the error in question. Is that under your control and could you force slicing somehow?

**Comment by RaulTrombin** [comment]
```
I'm using it like this:

export_cpy!(
    mod navigator {
        fn set_neopixel(array: &[u8; 3]) {
                 "content"
        }
  }
);
```
full code on:
[this commit](https://github.com/bluerobotics/navigator-lib/pull/25/commits/763f788160cfc45a8ef45e8def4b04b04a0d4f3e)

This macro will apply the transformations pointed here:
[cpy-rs-lib](https://github.com/patrickelectric/cpy-rs/blob/master/src/lib.rs)



for C it exported:
void set_neopixel(const uint8_t (*array)[3]);
and for python (with pyo3) the error.

If I change my input to be without a reference:
         fn set_neopixel(array: [u8; 3]) {
                 "content"
                 }
it works fine for the python(pyo3) side, but not give the C binding for the function as it says it is FFI unsafe.


@adamreichold  Do you mean try another type on entry instead &[ ]?

**Reply by adamreichold** [reply]
If you want to pass an array, then for PyO3 you should pass by value, i.e. `fn set_neopixel(array: [u8; 3])`.

For the C side, passing by value should be equivalent to passing a pointer, i.e. `void set_neopixel(const uint8_t* array)` because arrays always decay to pointers automatically, i.e. this should actually equivalent to `void set_neopixel(const uint8_t array[3])`. (The syntax you write above means an array of three pointers if I am not mistaken (which is then actually passed as a pointer to a pointer).)

**Reply by RaulTrombin** [reply]
Yes, it seems a language limitation issue, we did some new macro rules, on the right is the expansion.
![image](https://github.com/PyO3/pyo3/assets/80598030/962a9fd6-634a-4efb-8f17-0fa0ed09cfbe)

Now it keep working for cpp and work for for python.
